Sample message #2 HCI mapping updates

jimmyfagan opened the pull request:
This PR makes another batch of adjustments to mappings based on HCI's second sample message.

Test Steps:
1. Run integration tests
2. Use CLI commands to run manual conversion on the HCI sample message from Sharepoint
3. Run `./prime hl7data -c src/testIntegration/resources/datatests/HL7_to_FHIR_to_HL7/hci-expected.hl7 -s src/testIntegration/resources/datatests/HL7_to_FHIR_to_HL7/hci.hl7`

## Changes
- Remove default values for MSH.17/MSH.18 (this required updates to several test files, waiting to hear back from Victor about whether or not we need a receiver transform there for CA)
- Add a receiver transform to maintain original behavior for MSH.18 for SR->CA (MSH.17 was already being sent by SR)
- Add extensions for XCN->Practitioner.Name to support and preserve the different potential suffix types
- Update ExtensionAssigningAuthority to preserve HD.2 even when HD.3 is not set (this addresses an issue with PL.4, but as a short-term solution, does not fix the mapping for PL.4 altogether)


## Checklist

### Testing
- [x] Tested locally?
- [x] Ran `./prime test` or `./gradlew testSmoke` against local Docker ReportStream container?
- [x] Added tests?

## Linked Issues
- Addresses #11317

jimmyfagan commented:
Moving to draft temporarily in anticipation of incoming merge conflicts

arnejduranovic commented:
Looking at the AC of the associated ticket, it is not clear just by looking at the spreadsheet what actually got resolved and what didn't.

arnejduranovic reviewed (commented):
Looks good, with some comments/questions.

1. Can we update the spreadsheet once this is merged to make it clear what's left? Can we also describe what's left in this PR or ticket description?

I tested with the sample file in the sharepoint, and this is what the out HL7v2 diffed with the input looks like:

```
./prime fhirdata -i ~/Downloads/hci-test-files/hci-8-27-23-in_hl7.hl7 -o ~/Downloads/hci-test-files/arnej-out.hl7 --output-format HL7 -s /Users/arnej/Projects/prime-reportstream/prime-router/metadata/hl7_mapping/ORU_R01/ORU_R01-base.yml --diff-hl7-output TRUE
```

```
Wrote output to /Users/arnej/Downloads/hci-test-files/arnej-out.hl7
-------diff output
There were 48 differences between the input and output
Difference between messages at MSH(1).3 Differences: Difference in number of extra components. 
Input: ExtraComponents[] Output: ExtraComponents[Varies[PI]]
Difference between messages at MSH(1).8.1 Differences: Input had more repeating types for ST, input has 1 and output has 0
Difference between messages at MSH(1).17.1 Differences: Input had more repeating types for ID, input has 1 and output has 0
Difference between messages at MSH(1).18.1 Differences: Input had more repeating types for ID, input has 1 and output has 0
Difference between messages at SFT(1).5.1 Differences: Input had more repeating types for TX, input has 1 and output has 0
Difference between messages at PATIENT_RESULT(1)-PATIENT(1)-PID(1).13 Differences: Difference in number of components. Input: [PRN, Internet, noemail@prophasedx.com], Output: [noemail@prophasedx.com] 
Difference between messages at PATIENT_RESULT(1)-PATIENT(1)-PID(1).16.1 Differences: Input had more repeating types for CWE, input has 1 and output has 0
Difference between messages at PATIENT_RESULT(1)-PATIENT(1)-PID(1).33.1 Differences: Input had more repeating types for DTM, input has 1 and output has 0
Difference between messages at PATIENT_RESULT(1)-ORDER_OBSERVATION(1)-ORC(1).4.1 Differences: Output had more repeating types for EI, input has 0 and output has 1
Difference between messages at PATIENT_RESULT(1)-ORDER_OBSERVATION(1)-ORC(1).9 Differences: 20230815164300+0000, 20230815164300.
Difference between messages at PATIENT_RESULT(1)-ORDER_OBSERVATION(1)-ORC(1).12.4 Differences: Difference in number of extra components. 
Input: ExtraComponents[] Output: ExtraComponents[Varies[L]]
Difference between messages at PATIENT_RESULT(1)-ORDER_OBSERVATION(1)-ORC(1).23 Differences: Difference in number of components. Input: [PH, 1, 752, 2837, 752 2837], Output: [PH, 1, 7522837] 
Difference between messages at PATIENT_RESULT(1)-ORDER_OBSERVATION(1)-OBR(1).7 Differences: 20230815164300+0000, 20230815164300.
Difference between messages at PATIENT_RESULT(1)-ORDER_OBSERVATION(1)-OBR(1).16.4 Differences: Difference in number of extra components. 
Input: ExtraComponents[] Output: ExtraComponents[Varies[L]]
Difference between messages at PATIENT_RESULT(1)-ORDER_OBSERVATION(1)-OBR(1).24.1 Differences: Input had more repeating types for ID, input has 1 and output has 0
Difference between messages at PATIENT_RESULT(1)-ORDER_OBSERVATION(1)-OBSERVATION(1)-OBX(1).4.1 Differences: Input had more repeating types for ST, input has 1 and output has 0
Difference between messages at PATIENT_RESULT(1)-ORDER_OBSERVATION(1)-OBSERVATION(1)-OBX(1).5 Differences: Difference in type of field, class ca.uhn.hl7v2.model.GenericPrimitive, class ca.uhn.hl7v2.model.v27.datatype.NM.
Difference between messages at PATIENT_RESULT(1)-ORDER_OBSERVATION(1)-OBSERVATION(1)-OBX(1).14 Differences: 20230815164300+0000, 20230815164300.
Difference between messages at PATIENT_RESULT(1)-ORDER_OBSERVATION(1)-OBSERVATION(1)-NTE(1).2.1 Differences: Input had more repeating types for ID, input has 1 and output has 0
Difference between messages at PATIENT_RESULT(2)-ORDER_OBSERVATION(1)-ORC(1).4.1 Differences: Output had more repeating types for EI, input has 0 and output has 1
Difference between messages at PATIENT_RESULT(2)-ORDER_OBSERVATION(1)-ORC(1).9 Differences: 20230815164300+0000, 20230815164300.
Difference between messages at PATIENT_RESULT(2)-ORDER_OBSERVATION(1)-ORC(1).12.4 Differences: Difference in number of extra components. 
Input: ExtraComponents[] Output: ExtraComponents[Varies[L]]
Difference between messages at PATIENT_RESULT(2)-ORDER_OBSERVATION(1)-ORC(1).23 Differences: Difference in number of components. Input: [PH, 1, 752, 2837, 752 2837], Output: [PH, 1, 7522837] 
Difference between messages at PATIENT_RESULT(2)-ORDER_OBSERVATION(1)-OBR(1).7 Differences: 20230815164300+0000, 20230815164300.
Difference between messages at PATIENT_RESULT(2)-ORDER_OBSERVATION(1)-OBR(1).16.4 Differences: Difference in number of extra components. 
Input: ExtraComponents[] Output: ExtraComponents[Varies[L]]
Difference between messages at PATIENT_RESULT(2)-ORDER_OBSERVATION(1)-OBR(1).24.1 Differences: Input had more repeating types for ID, input has 1 and output has 0
Difference between messages at PATIENT_RESULT(2)-ORDER_OBSERVATION(1)-OBSERVATION(1)-OBX(1).4.1 Differences: Input had more repeating types for ST, input has 1 and output has 0
Difference between messages at PATIENT_RESULT(2)-ORDER_OBSERVATION(1)-OBSERVATION(1)-OBX(1).5 Differences: Difference in type of field, class ca.uhn.hl7v2.model.GenericPrimitive, class ca.uhn.hl7v2.model.v27.datatype.ST.
Difference between messages at PATIENT_RESULT(2)-ORDER_OBSERVATION(1)-OBSERVATION(1)-OBX(1).7.1 Differences: Input had more repeating types for ST, input has 1 and output has 0
Difference between messages at PATIENT_RESULT(2)-ORDER_OBSERVATION(1)-OBSERVATION(1)-OBX(1).14 Differences: 20230815164300+0000, 20230815164300.
Difference between messages at PATIENT_RESULT(3)-ORDER_OBSERVATION(1)-OBR(1).7 Differences: 20230815164300+0000, 20230815164300.
Difference between messages at PATIENT_RESULT(3)-ORDER_OBSERVATION(1)-OBR(1).16.4 Differences: Difference in number of extra components. 
Input: ExtraComponents[] Output: ExtraComponents[Varies[L]]
Difference between messages at PATIENT_RESULT(3)-ORDER_OBSERVATION(1)-OBR(1).24.1 Differences: Input had more repeating types for ID, input has 1 and output has 0
Difference between messages at PATIENT_RESULT(3)-ORDER_OBSERVATION(1)-OBSERVATION(1)-OBX(1).4.1 Differences: Input had more repeating types for ST, input has 1 and output has 0
Difference between messages at PATIENT_RESULT(3)-ORDER_OBSERVATION(1)-OBSERVATION(1)-OBX(1).5 Differences: Difference in type of field, class ca.uhn.hl7v2.model.GenericPrimitive, class ca.uhn.hl7v2.model.v27.datatype.ST.
Difference between messages at PATIENT_RESULT(3)-ORDER_OBSERVATION(1)-OBSERVATION(1)-OBX(1).7.1 Differences: Input had more repeating types for ST, input has 1 and output has 0
Difference between messages at PATIENT_RESULT(3)-ORDER_OBSERVATION(1)-OBSERVATION(1)-OBX(1).14 Differences: 20230815164300+0000, 20230815164300.
Difference between messages at PATIENT_RESULT(4)-ORDER_OBSERVATION(1)-OBR(1).7 Differences: 20230815164300+0000, 20230815164300.
Difference between messages at PATIENT_RESULT(4)-ORDER_OBSERVATION(1)-OBR(1).16.4 Differences: Difference in number of extra components. 
Input: ExtraComponents[] Output: ExtraComponents[Varies[L]]
Difference between messages at PATIENT_RESULT(4)-ORDER_OBSERVATION(1)-OBR(1).24.1 Differences: Input had more repeating types for ID, input has 1 and output has 0
Difference between messages at PATIENT_RESULT(4)-ORDER_OBSERVATION(1)-OBSERVATION(1)-OBX(1).4.1 Differences: Input had more repeating types for ST, input has 1 and output has 0
Difference between messages at PATIENT_RESULT(4)-ORDER_OBSERVATION(1)-OBSERVATION(1)-OBX(1).5 Differences: Difference in type of field, class ca.uhn.hl7v2.model.GenericPrimitive, class ca.uhn.hl7v2.model.v27.datatype.ST.
Difference between messages at PATIENT_RESULT(4)-ORDER_OBSERVATION(1)-OBSERVATION(1)-OBX(1).7.1 Differences: Input had more repeating types for ST, input has 1 and output has 0
Difference between messages at PATIENT_RESULT(4)-ORDER_OBSERVATION(1)-OBSERVATION(1)-OBX(1).14 Differences: 20230815164300+0000, 20230815164300.
Difference between messages at PATIENT_RESULT(4)-ORDER_OBSERVATION(1)-SPECIMEN(1)-SPM(1).11.1 Differences: Input had more repeating types for CWE, input has 1 and output has 0
Difference between messages at PATIENT_RESULT(4)-ORDER_OBSERVATION(1)-SPECIMEN(1)-SPM(1).14.1 Differences: Input had more repeating types for ST, input has 1 and output has 0
Difference between messages at PATIENT_RESULT(4)-ORDER_OBSERVATION(1)-SPECIMEN(1)-SPM(1).17.1 Differences: 20230815164300+0000, 20230815164300.
Difference between messages at PATIENT_RESULT(4)-ORDER_OBSERVATION(1)-SPECIMEN(1)-SPM(1).21.1 Differences: Input had more repeating types for CWE, input has 1 and output has 0
```

jimmyfagan commented:
It looks like the reported vs actual differences with the difftool are pretty significant, so trying to do some debugging on that, and maybe include some difftool fixes as part of this PR.

jimmyfagan commented on prime-router/src/testIntegration/resources/datatests/FHIR_to_HL7/sample_OML_20230831-0001.hl7:
```diff
@@ -1,4 +1,4 @@
-MSH|^~\&|Epic^1.2.840.114350.1.13.145.2.7.2.695071^ISO|Centracare^centracare.com^DNS|NATUS^natus.health.state.mn.us^DNS|MN Public Health Lab^2.16.840.1.114222.4.1.10080^ISO|20230506052916-0500||OML^O21^OML_O21|31808297|P|2.5.1|||AL|AL|USA|UNICODE UTF-8|||LAB_PRU_COMPONENT^^2.16.840.1.113883.9.82^ISO~LAB_TO_COMPONENT^^2.16.840.1.113883.9.22^ISO
+MSH|^~\&|Epic^1.2.840.114350.1.13.145.2.7.2.695071^ISO|Centracare^centracare.com^DNS|NATUS^natus.health.state.mn.us^DNS|MN Public Health Lab^2.16.840.1.114222.4.1.10080^ISO|20230506052916-0500||OML^O21^OML_O21|31808297|P|2.5.1|||AL|AL|||||LAB_PRU_COMPONENT^^2.16.840.1.113883.9.82^ISO~LAB_TO_COMPONENT^^2.16.840.1.113883.9.22^ISO
```
For now, `UNICODE UTF-8` is included in CA's receiver transform (until we can confirm with them that it can be removed). So tests like sample_AOE that use the CA receiver transform keep it, while in any other cases where it wasn't needed, it was removed.

jimmyfagan commented on prime-router/metadata/fhir_mapping/hl7/resource/Practitioner.yml:
```diff
@@ -18,7 +18,7 @@ identifier_1:
     typeSegment: XCN.13
 
 name:
-  specs: XCN | XCN | CNN | NDL | PPN
```
XPN might make sense here, but there is no XPN->Practitioner mapping, so I don't think we'd be in that scenario (XPN would normally go directly from the segment into HumanName like in the case of PID.5), so I'll leave it out. Out of curiousity I looked into it a bit more, and the duplicate is in the base library 😬

jimmyfagan commented:
Needed to add another pretty big commit to allow these changes to not break FHIR tests (due to the new strict FHIR comparison). Made another few edits to FHIR mappings to add conditions and prevent empty resources/values.

Instead of blocking this on the diff tool being updated, I would recommend checking the in/out files on sharepoint. The output file has been updated based on the latest in this branch. The differences are:

- Formatting; trimmed empty fields or sub-fields
- Timestamps; added timezones where they were missing
- Adjustments to XTNs; adjusted badly formatted phone number, added data to email XTN (this overlaps with Brick's story, we might need a small follow-up after that's merged)
- Unsupported data; OBR.16.9.4 is lost, but that is not part of the spec for 2.5.1 or 2.7

jimmyfagan commented on prime-router/metadata/hl7_mapping/common/datatype/xcn-contact.yml:
```diff
@@ -18,14 +18,28 @@ elements:
     hl7Spec: [ '%{hl7XCNFieldPath}-4' ]
 
-  - name: xcn-suffix
+  - name: xcn-suffix-from-value
+    condition: >
+      %resource.name.extension(%`rsext-name-suffix`).exists().not() and 
+      %resource.name.extension(%`rsext-name-suffix-degree`).exists().not() and 
+      %resource.name.extension(%`rsext-name-suffix-professional`).exists().not()
     value: [ '%resource.name.suffix.toString()' ]
     hl7Spec: [ '%{hl7XCNFieldPath}-5' ]
```
I was trying to remove this element based on out conversation this morning, but it broke a handful of translation tests.. Not sure how to deal with it, but I'm leaning towards adding the extension to our sample fhir. Don't think there's an across-the-board sender transform we can use here since I'm sure the type of suffix depends on who the person is and what specific suffices we're dealing with.

jimmyfagan replied:
I'll maybe bring it up as another 17th tomorrow, and this'll just be slightly unresolved til then.

jimmyfagan replied:
Ended up handling this in [this commit](https://github.com/CDCgov/prime-reportstream/pull/11850/commits/daeb088120aae569baae65ea54229bbac0b05b4a). Instead of handling it with sender transforms, I just edited the FHIR samples (since they don't correspond to current actual senders).

brick-green commented on prime-router/metadata/fhir_mapping/hl7/datatype/HumanName.yml:
```diff
@@ -1,56 +1,98 @@
+        valueString:
+          type: STRING
+          valueOf: $professionalSuffix
+          expressionType: HL7Spec
```
Why is `expressionType: HL7Spec` needed on these?

brick-green commented on prime-router/metadata/fhir_mapping/hl7/datatype/HumanName.yml:
```diff
@@ -1,56 +1,95 @@
+extension:
+  generateList: true
+  expressionType: nested
+  condition: $suffix NOT_NULL || $suffixDegree NOT_NULL || $professionalSuffix NOT_NULL
```
What do you think about have the conditions be applied at the expressions level rather than the top level?

jimmyfagan replied:
I love it, I thought this was required to prevent an empty `name` from appearing, but it seems like it's working actually, just committed the fix.

brick-green reviewed (approved):
looks good!

jimmyfagan commented on prime-router/src/testIntegration/resources/datatests/FHIR_to_HL7/sample_SR_2_20231024-0002.hl7:
```diff
@@ -1 +1,12 @@
\ No newline at end of file
+MSH|^~\&|CDC PRIME - Atlanta^2.16.840.1.114222.4.1.237821^ISO|CDC PRIME REPORTSTREAM^10DRPTSTRM^CLIA|ECLRS|NYSDOH|20231018222022+0000||ORU^R01^ORU_R01|5f0d29c2-4181-4a1e-8774-b88338602681|T|2.5.1|||||USA|UNICODE UTF-8|||PHLabReport-NoAck^ELR_Receiver^2.16.840.1.113883.9.11^ISO
+SFT|Centers for Disease Control and Prevention|0.2-SNAPSHOT|PRIME ReportStream|0.2-SNAPSHOT||20231018210603+0000
+PID|1||b023143c-b30d-4567-b651-be966ebd5532^^^Carl Potts&12D3456789&CLIA^PI^Carl Potts&12D3456789&CLIA||West^Theodora^Stacy^^^^L||19660304|F||ASKU^refused^^^^^2.5.1^^refused|40405 Lionel Lakes^^New Savannah^NY^10001^USA||^PRN^PH^^1^410^7542890^^^^^(410) 754 2890~^NET^Internet^Theodora4@yahoo.com|||||||||N^Not Hispanic or Latino^HL70189^^^^2.9^^Not Hispanic or Latino
```
The only real change here was `WPN`->`NET` in PID.13 based on @victor-chaparro 's advice.

brick-green commented on prime-router/metadata/fhir_mapping/hl7/datatypes/XTN/ContactPoint.yml:
```diff
@@ -25,50 +25,51 @@ value_3:
   value: "home"
   vars:
-    useCode: String, XTN.2
+    hl7Use: String, XTN.2
```
nit: maybe this should be `xtnUse` instead of `hl7Use`?

jimmyfagan replied:
Yea good idea, will add that.

brick-green commented on prime-router/metadata/fhir_mapping/hl7/codesystem/ExtensionUrlMapping.yml:
```diff
@@ -337,4 +337,13 @@
 - id: "hl7-use"
-  url: "https://reportstream.cdc.gov/fhir/StructureDefinition/hl7-use"
\ No newline at end of file
+  url: "https://reportstream.cdc.gov/fhir/StructureDefinition/hl7-use"
```
Looks like this one can be removed since it's not part of the schema anymore

arnejduranovic commented on prime-router/metadata/fhir_mapping/hl7/codesystem/ExtensionUrlMapping.yml:
```diff
@@ -337,4 +337,13 @@
+# Name suffix degree
+- id: "name-suffix-degree"
+  url: "https://reportstream.cdc.gov/fhir/StructureDefinition/name-suffix-degree"
+# Name suffix
```
```suggestion
# Name suffix professional
```

brick-green commented on prime-router/metadata/fhir_mapping/hl7/secondary/Contact.yml:
```diff
@@ -13,7 +13,7 @@ name:
 telecom:
   condition: $contactPointXTN NOT_NULL
-  valueOf: datatype/ContactPoint
+  valueOf: datatypes/XTN/ContactPoint
```
This change makes me a little nervous along with the change in order.yml since it covers ORC-23 which is an important field for STLTs. It's fine for HL7 senders but how will it work for SR on the outbound to HL7?

jimmyfagan replied:
We discussed this in a huddle, it's not super clear, but this is covered as part of the sender transform for SR, so this concern is addressed.

arnejduranovic commented on prime-router/src/main/kotlin/fhirengine/translation/hl7/FhirToHl7Converter.kt:
```diff
@@ -87,10 +87,6 @@ class FhirToHl7Converter(
         val message = HL7Utils.getMessageInstance(schemaRef.hl7Class!!)
 
         terser = Terser(message)
-        val dupes = schemaRef.duplicateElements
```
Why was this code here in the first place? Do you we have tests that verify what happens when there are duplicate elements?

jimmyfagan replied:
I'll go back and edit the existing tests rather than remove them entirely.

arnejduranovic reviewed (commented):
Changes look good to me overall. I left a comment I'd like some feedback on before approving. I tested HL7 -> FHIR -> HL7 with the latest HCI message on your branch and only differences were we were removing empty fields. I think this is fine for now but would like to discuss how we are handling this with the new mappings as a 17th.

jimmyfagan commented on prime-router/metadata/fhir_transforms/senders/SimpleReport/simple-report-sender-transform.yml:
```diff
@@ -1,6 +1,6 @@
 constants:
-  serviceRequest:  'Bundle.entry.resource.ofType(ServiceRequest)'
+  serviceRequest: 'Bundle.entry.resource.ofType(ServiceRequest)'
   diagnosticReport: 'Bundle.entry.resource.ofType(DiagnosticReport)'
```
@victor-chaparro Can you confirm that we never expect to have more than 1 ServiceRequest or DiagnosticReport or Patient from SR? I think these constants not having indices means we'd only apply the sender transforms to the first instance of each of these; if we had multiple, we'd need to split things out into sub-schemas.

arnejduranovic reviewed (approved):
Approving after our chat, we discussed reworking the unit tests instead of deleting them but otherwise everything looks good pending any comments from Engagement on SR changes.
